cartographer_ros/cartographer_ros/set_initpose_main.cc

```diff
@@ -0,0 +1,181 @@
+                               itr->data.pose.translation().y(),
+                               itr->data.pose.translation().z());
+
+    // calculate the distance between submap origin and initial pose
+    tf2::Vector3 delta_pos = submap_origin - map_tf.getOrigin();
+    delta_pos.setZ(0);  // only check horinzontal location
+
+    // find the best submap, which is closest to the initial pose
+    // and assign the height (i.e. z value) of this submap to the initial pose
+    if (delta_pos.length() < min_dist) {
+      map_tf.getOrigin().setZ(submap_origin.z());
+      min_dist = delta_pos.length();
+    }
+  }
+
+  tf2::Transform relative_initpose_tf = traj_ref_tf.inverse() * map_tf;
+
+  // set the start trajectory service call
+  ::ros::ServiceClient client_start_trajectroy =
+      nh.serviceClient<cartographer_ros_msgs::StartTrajectory>(
+          cartographer_ros::kStartTrajectoryServiceName);
+  cartographer_ros_msgs::StartTrajectory srv_start_trajectroy;
+  srv_start_trajectroy.request.configuration_directory =
+      FLAGS_configuration_directory;
+  srv_start_trajectroy.request.configuration_basename =
+      FLAGS_configuration_basename;
+
+  srv_start_trajectroy.request.relative_to_trajectory_id =
+      0;  // frozen trajectory
+  srv_start_trajectroy.request.use_initial_pose = true;
+  tf2::toMsg(relative_initpose_tf, srv_start_trajectroy.request.initial_pose);
+
+  if (!client_start_trajectroy.call(srv_start_trajectroy)) {
+    LOG(ERROR) << "Failed to call "
+               << cartographer_ros::kStartTrajectoryServiceName << ".";
+  }
+  if (srv_start_trajectroy.response.status.code !=
+      cartographer_ros_msgs::StatusCode::OK) {
+    LOG(ERROR) << "Error starting trajectory - message: '"
+               << srv_start_trajectroy.response.status.message
+               << "' (status code: "
+               << std::to_string(srv_start_trajectroy.response.status.code)
+               << ").";
+  }
+}
+
+int main(int argc, char** argv) {
+  google::InitGoogleLogging(argv[0]);
+
+  google::SetUsageMessage(
+      "\n\n"
+      "Convenience tool around the start_trajectory service. This takes a Lua "
+      "file that is accepted by the node as well and starts a new trajectory "
+      "using its settings.\n");
+
+  google::ParseCommandLineFlags(&argc, &argv, true);
+
+  CHECK(!FLAGS_configuration_basename.empty())
+      << "-configuration_basename is missing.";
+
+  CHECK(!FLAGS_configuration_directory.empty())
+      << "-configuration_directory is missing.";
+
+  // load pbstream
+  ::cartographer::io::ProtoStreamReader reader(FLAGS_load_state_filename);
+  cartographer_ros::NodeOptions node_options;
+  std::tie(node_options, std::ignore) = cartographer_ros::LoadOptions(
+      FLAGS_configuration_directory, FLAGS_configuration_basename);
+  map_builder_ = absl::make_unique<cartographer::mapping::MapBuilder>(
+      node_options.map_builder_options);
+  map_builder_->LoadState(&reader, true);
```

Comment:
map build is used to get the frozen trajectory information:
https://github.com/googlecartographer/cartographer_ros/blob/5b09b08daf6a6f1f51296f44e11f8c1f5ba48d40/cartographer_ros/cartographer_ros/set_initpose_main.cc#L109

The trajectory information is required here, since the initial pose of the new trajectory should be provided w.r.t the reference trajectory (i.e. the frozen trajectory). Is it right?